components/ContractViewer/ViewerUtils.ts

```diff
@@ -0,0 +1,120 @@
+// eslint-disable-next-line prettier/prettier
+import { type AbiParameter, type AbiInternalType } from 'abitype'
+import { createPublicClient, http } from 'viem'
+import { mainnet } from 'viem/chains'
+
+export const rpc = createPublicClient({
+  chain: mainnet,
+  transport: http('https://eth.merkle.io/'),
+})
+
+// redefined here because maybe there is a bug in abitypes,
+// since components is not recognized
+export type AbiComponent = AbiParameter & { 
+  components: AbiComponent[]
+  internalType?: AbiInternalType
+}
+
+export const getBadgeColor = (str: string) => {
+  // tailwind wants the entire class name at compile time, so don't build it dynamically
+  const colors = [
+    'bg-red-300',
+    'bg-green-300',
+    'bg-blue-300',
+    'bg-purple-300',
+    'bg-pink-300',
+  ]
+  const hash = str.charCodeAt(0) + str.charCodeAt(str.length - 1)
+  return colors[hash % colors.length]
+}
+
+export function getTypePrettyName(type: any): string {
+  if (typeof type == 'string') {
+    return type
+  }
+
+  // TODO: doesn't always work (eg. storage mapping as param? see aave pools for example)
+  if (type.type == 'ArrayTypeName') {
+    // TODO: support fixed array length
+    return getTypePrettyName(type.baseTypeName) + '[]'
+  } else if (type.type == 'Mapping') {
+    return (
+      'mapping (' +
+      getTypePrettyName(type.keyType) +
+      ' => ' +
+      getTypePrettyName(type.valueType) +
+      ')'
+    )
+  }
+
+  return type?.name || type?.namePath || ''
+}
+
+export function spaceBetween(str: string, pad = 16) {

```

Comment:
Maybe we should have a test for this function ? 

It's a bit complex so having some test would make it easier to simplify in the future

Reply:
that is a conversation for another day...